<a href="https://colab.research.google.com/github/adas754/python_revision-from-day1/blob/main/python_140.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Program 1
# Write a Python program to print "Hello Python"

In [ ]:
print('Hello Python')

Hello Python


#  Program 2 Write a Python program to do arithmetical operations addition and division.

In [ ]:
def add_div(a,b):
  c=a+b
  d=a/b
  return c,d
add_div(25,5)

(30, 5.0)

# Program 3 Write a Python program to find the area of a triangle.

In [ ]:
def area(h,b):
  return 0.5*h*b
area(10,15)

75.0

# Program 4 Write a Python program to swap two variables.

In [ ]:
def swap(a,b):
  a=a+b
  b=a-b
  a=a-b
  return a,b
a,b=swap(34,67)
print(a,b)

67 34


# Program 5
# Write a Python program to generate a random number.

In [ ]:
import random
random.randint(1,100)

25

# Program 6
# Write a Python program to convert kilometers to miles

In [ ]:
def km_mile(km):
  miles=km* 0.621371
  return miles
km_mile(100)


62.137100000000004

# Program 7
# Write a Python program to convert Celsius to Fahrenheit.

In [ ]:
def cel_far(c):
  f=(c*9/5)+32
  return f
cel_far(37)

98.6

# Program 8
# Write a Python program to display calendar.

In [ ]:
import calendar
year = int(input("Enter year: "))
month = int(input("Enter month: "))
cal = calendar.month(year, month)
print(cal)


Enter year: 2023
Enter month: 11
   November 2023
Mo Tu We Th Fr Sa Su
       1  2  3  4  5
 6  7  8  9 10 11 12
13 14 15 16 17 18 19
20 21 22 23 24 25 26
27 28 29 30



# Program 9
# Write a Python program to solve quadratic equation.

In [ ]:
import cmath
a=int(input())
b=int(input())
c=int(input())
d=(b**2-4*a*c)**0.5
root1=(-b+d)/(2*a)
root2=(-b-d)/(2*a)
print(root1,root2)


1
4
8
(-1.9999999999999998+2j) (-2-2j)


# Program 10
# Write a Python program to swap two variables without temp variable.

In [ ]:
def swap_num(a,b):
  a=a+b
  b=a-b
  a=a-b
  return a,b
swap_num(56,57)

(57, 56)

# Program 11 Write a Python Program to Check if a Number is Positive, Negative or Zero.

In [ ]:
def pos_num(n):
  if n>0:
    print('positive')
  elif n==0:
    print('zero')
  else:
    print('negative')
pos_num(56)


positive


# Program 12 Write a Python Program to Check if a Number is Odd or Even

In [ ]:
def eve_odd(n):
  if n%2==0:
    print('even')
  else:
    print('odd')
eve_odd(31)

odd


# Program 13 Write a Python Program to Check Leap Year.

In [ ]:
def leap_yr(yr):
  if(yr%4==0 and yr%100!=0 or yr%400==0):
    print('year is leap year')
  else:
    print('not leap year')
leap_yr(2024)

year is leap year


# Program 14
# Write a Python Program to Check Prime Number.


In [ ]:
def check_prime(n):
  if n<=1:
    return False
  for i in range(2,int(n**0.5)+1):
    if n%i==0:
      return False
  return True
check_prime(19)

True

# Program 15
# Write a Python Program to Print all Prime Numbers in an Interval of 1-10.


In [ ]:
def is_prime(n):
  if n<=1:
    return False
  for i in range(2,int(n**0.5)+1):
    if n%i==0:
      return False
  return True
def print_prime(start,end):
  for i in range(start,end+1):
    if is_prime(i):
      print(i)
print_prime(1,10)

2
3
5
7


# Program 16
# Write a Python Program to Find the Factorial of a Number.


In [ ]:
def fact_number(n):
  if n==0 or n==1:
    return 1
  else:
    return n*fact_number(n-1)
fact_number(5)

120

# Program 17
# Write a Python Program to Display the multiplication Table.


In [ ]:
n=int(input())
for i in range(1,11):
  print(f'{n}x{i}={n*i}')

10
10x1=10
10x2=20
10x3=30
10x4=40
10x5=50
10x6=60
10x7=70
10x8=80
10x9=90
10x10=100


# Program 18
# Write a Python Program to Print the Fibonacci sequence.

In [ ]:
def fib_num(n):
  if n<=1:
    return n
  else:
    return (fib_num(n-1)+fib_num(n-2))
for i in range(10):
  print(fib_num(i))

0
1
1
2
3
5
8
13
21
34


# Program 19
# Write a Python Program to Check Armstrong Number?

In [ ]:
def arm_num(num):
  num_str=str(num)
  digit_num=len(num_str)
  sum=0
  for digit in num_str:
    sum+=int(digit)**digit_num
  if sum==num:
    return True
  else:
    return False
arm_num(9474)

True

# Program 20
# Write a Python Program to Find Armstrong Number in an Interval.

In [ ]:
def arm_num(num):
  num_str=str(num)
  digit_num=len(num_str)
  sum=0
  for digit in num_str:
    sum+=int(digit)**digit_num
  if sum==num:
    return True
  else:
    return False
def print_arm(start,end):
  for i in range(start,end+1):
    if arm_num(i):
      print(i)
print_arm(10,1000)

153
370
371
407


# Program 21
# Write a Python Program to Find the Sum of Natural Numbers.

In [ ]:
def son(n):
  s=0
  for i in range(1,n+1):
    s+=i
  return s
son(10)

55

# Program 22
# Write a Python Program to Find LCM.

In [ ]:
def compute_lcm(x,y):
  if x>y:
    greater=x
  else:
    greater=y
  while True:
    if((greater%x==0)and(greater%y==0)):
      lcm=greater
      break
    greater+=1
  return lcm
compute_lcm(54,24)

216

# Program 23
# Write a Python Program to Find HCF.


In [ ]:
def compute_hcf(x,y):
  if x>y:
    smaller=y
  else:
    smaller=x
  for i in range(1,smaller+1):
    if((x%i==0)and(y%i==0)):
      hcf=i
  return hcf
compute_hcf(54,24)

6

# Program 24
# Write a Python Program to Convert Decimal to Binary, Octal and Hexadecimal.

In [ ]:
dec_num=int(input("Enter decimal "))
print(bin(dec_num),'in binary')
print(oct(dec_num),'in octal')
print(hex(dec_num),'in hexadecimal')

Enter decimal 27
0b11011 in binary
0o33 in octal
0x1b in hexadecimal


# Program 25
# Write a Python Program To Find ASCII value of a character.


In [ ]:
char=str(input('Enter the character :'))
print('asci value of : ',ord(char))

Enter the character :p
asci value of :  112


# Program 26
# Write a Python Program to Make a Simple Calculator with 4 basic mathematical operations.

In [ ]:
def calculate(a,b,operator):
  if operator=='+':
    return a+b
  elif operator=='-':
    return a-b
  elif operator=='*':
    return a*b
  elif operator=='/':
    return a/b
  else:
    return 'invalid operator'
def main():
  while True:
    a=float(input('Enter 1st number :'))
    b=float(input('Enter 2nd number :'))
    operator=input('Enter operator(+,-,*,/):')
    result=calculate(a,b,operator)
    print(f"result:{result}")
    if input('do you want to continue(y/n) ?')=='n':
      break
main()




Enter 1st number :5
Enter 2nd number :4
Enter operator(+,-,*,/):+
result:9.0
do you want to continue(y/n) ?y
Enter 1st number :5
Enter 2nd number :5
Enter operator(+,-,*,/):-
result:0.0
do you want to continue(y/n) ?y
Enter 1st number :5
Enter 2nd number :4
Enter operator(+,-,*,/):*
result:20.0
do you want to continue(y/n) ?y
Enter 1st number :40
Enter 2nd number :5
Enter operator(+,-,*,/):/
result:8.0
do you want to continue(y/n) ?n


# Program 27
# Write a Python Program to Display Fibonacci Sequence Using Recursion.


In [ ]:
def fib(n):
  if n<=1:
    return n
  else:
    return fib(n-1)+fib(n-2)
for i in range(8):
  print(fib(i))

0
1
1
2
3
5
8
13


# Program 28
# Write a Python Program to Find Factorial of Number Using Recursion.

In [ ]:
def fact(n):
  if n==0 or n==1:
    return n
  else:
    return n*fact(n-1)
fact(5)

120

# Program 29 Write a Python Program to calculate your Body Mass Index.

In [ ]:
def bodymassindex(h,w):
  return round((w/h**2),2)
h=float(input('Enter height '))
w=float(input('Enter weight '))
print('welcome BMI calculator ')
bmi=bodymassindex(h,w)
print('your bmi is',bmi)
if bmi <=18.5:
  print('you are underweight ')
elif 18.5<bmi<=24.9:
  print('your wight is normal ')
else:
  print('you are obese')

Enter height 18
Enter weight 70
welcome BMI calculator 
your bmi is 0.22
you are underweight 


# Program 30 Write a Python Program to calculate the natural logarithm of any number.


In [ ]:
import math
num=float(input('Enter number '))
if num<=0:
  print('please enter a positive number ')
else:
  result=math.log(num)
  print(result)

Enter number 1.4
0.3364722366212129


# Program 31
# Write a Python Program for cube sum of first n natural numbers?

In [ ]:
def natural(n):
  s=0
  for i in range(n+1):
    s=s+(i**3)
  return s
natural(7)

784

# Program 32
#Write a Python Program to find sum of array.


In [ ]:
def soa(arr):
  total=0
  for i in arr:
    total+=i
  return total
arr=[1,2,3]
soa(arr)

6

# Program 33
# Write a Python Program to find largest element in an array

In [ ]:
def find_largest(arr):
  max_val=arr[0]
  for i in arr:
    max_val=i if i>max_val else max_val
  return max_val
arr=[67,78,89]
find_largest(arr)

89

# Program 34
# Write a Python Program for array rotation

In [ ]:
n = [1,2,3,4,5]
rotate = 2
rotate = rotate % len(n)
ans = n[rotate:] + n[:rotate]
print(ans)

[3, 4, 5, 1, 2]


# Program 35
# Write a Python Program to Split the array and add the first part to the end?

In [ ]:
arr=[1,2,3,4,5]
k=3
new_arr=arr[k:]+arr[:k]
new_arr

[4, 5, 1, 2, 3]

# Program 36
# Write a Python Program to check if given array is Monotonic

In [ ]:
def is_monotonic(arr):
  increasing=decreasing=True
  for i in range(1,len(arr)):
    if arr[i]>arr[i-1]:
      decreasing=False
    elif arr[i]<arr[i-1]:
      increasing=False
  return increasing or decreasing
array1 = [1, 2, 3, 4, 5]
array2 = [5, 4, 3, 2, 1]
array3 = [1, 3, 2, 5, 7]
print(is_monotonic(array1))
print(is_monotonic(array2))
print(is_monotonic(array3))

True
True
False


# Program 37
# Write a Python Program to Add Two Matrices.


In [ ]:
def add_matrices(matrix1,matrix2):
  result=[]
  for i in range(len(matrix1)):
    row=[]
    for j in range(len(matrix1[0])):
      row.append(matrix1[i][j]+matrix2[i][j])
    result.append(row)
  return result

matrix1 = [[1, 2, 3], [4, 5, 6]]
matrix2 = [[7, 8, 9], [10, 11, 12]]
print(add_matrices(matrix1, matrix2))


[[8, 10, 12], [14, 16, 18]]


# Program 38
# Write a Python Program to Multiply Two Matrices.


In [ ]:
def multiply_matrices(matrix1,matrix2):
  result=[[0 for j in range(len(matrix1[0]))] for i in range(len(matrix1))]
  for i in range(len(matrix1)):
    for j in range(len(matrix2[0])):
      for k in range(len(matrix2)):
        result[i][j]+=matrix1[i][k]*matrix2[k][j]
  return result
matrix1 = [[1, 2], [3,4]]
matrix2 = [[5,6], [7,8]]
print(multiply_matrices(matrix1, matrix2))

[[19, 22], [43, 50]]


# Program 39
# Write a Python Program to Transpose a Matrix.

In [ ]:
def transpose(matrix):
  result=[[0 for i in range(len(matrix))]for j in range(len(matrix[0]))]
  for i in range(len(matrix)):
    for j in range(len(matrix[0])):
      result[j][i]=matrix[i][j]
  return result
matrix = [[1, 2, 3], [4, 5, 6]]
print(transpose(matrix))

[[1, 4], [2, 5], [3, 6]]


# Program 40
# Write a Python Program to Sort Words in Alphabetic Order.


In [ ]:
def sort_words(words):
  for i in range(len(words)):
    for j in range(i+1,len(words)):
      if words[i]>words[j]:
        temp=words[i]
        words[i]=words[j]
        words[j]=temp
  return words
words = ["red", "blue", "green", "yellow", "orange", "purple",'z']
print(sort_words(words))

['blue', 'green', 'orange', 'purple', 'red', 'yellow', 'z']


# Program 41
# Write a Python Program to Remove Punctuation From a String.

In [ ]:
punctuations = '''!()-[]{};:'"\,<>./?@#$%^&*_~'''
def remove_punctuation(string):
  no_punct=''
  for chr in string:
    if chr not in punctuations:
      no_punct+=chr
  return no_punct
my_str='hi@python!'
remove_punctuation(my_str)

'hipython'

# Program 42
#list to string

In [ ]:
l=['i','am','munai']
l=''.join(l)
l

'iammunai'

# Program 43
# Write a Python program to check if the given number is a Disarium Number.

In [ ]:
def is_disarium(num):
  num_str=str(num)
  len_str=len(num_str)
  sum=0
  for i in range(len(num_str)):
    sum+=int(num_str[i])**(i+1)
  return num==sum
is_disarium(89)

True

## Program 44
# Write a Python program to print all disarium numbers between 1 to 100.

In [ ]:
def is_disarium(num):
  num_str = str(num)
  len_str = len(num_str)
  sum = 0
  for i in range(len_str):
    sum += int(num_str[i]) ** (i + 1)
  return num == sum
disarium_numbers = [num for num in range(1, 101) if is_disarium(num)]
for num in disarium_numbers:
 print(num)


1
2
3
4
5
6
7
8
9
89


# Program 45
# Write a Python program to check if the given number is Happy Number

In [ ]:
def is_happy_number(num):
  seen=set()
  while num!=1 and num not in seen:
    seen.add(num)
    num=sum(int(i)**2 for i in str(num))
  return num==1
is_happy_number(23)

True

# Program 46
# Write a Python program to print all happy numbers between 1 and 100.


In [ ]:
def is_happy_number(num):
  seen=set()
  while num!=1 and num not in seen:
    seen.add(num)
    num=sum(int(i)**2 for i in str(num))
  return num==1
happy_numbers=[]
for num in range(1,101):
  if is_happy_number(num):
    happy_numbers.append(num)
print(happy_numbers)


[1, 7, 10, 13, 19, 23, 28, 31, 32, 44, 49, 68, 70, 79, 82, 86, 91, 94, 97, 100]


# Program 47
# Write a Python program to determine whether the given number is a Harshad Number.


In [ ]:
def is_harshad(num):
  digit_sum=sum(int(i) for i in str(num))
  return num%digit_sum==0
is_harshad(42)

True

# Program 48
# Write a Python program to print all pronic numbers between 1 and 100.

In [ ]:
def is_pronic(num):
  for n in range(1,int(num**0.5)+1):
    if n*(n+1)==num:
      return True
  return False
for i in range(1,101):
  if is_pronic(i):
    print(i)

2
6
12
20
30
42
56
72
90


# Program 49 Write a Python program to sum all numbers in the list.


In [ ]:
numbers=[7,9,2]
sum=0
for i in (numbers):
  sum+=i
print(sum)

18


# Program 50
# Write a Python program to Multiply all numbers in the list.


In [ ]:
num=[9,3,5]
mul=1
for i in num:
  mul*=i
print(mul)

135


# Program 51
# Write a Python program to find smallest number in a list.


In [ ]:
num=[6,-8,8,9]
min=num[0]
for i in num:
  if i<min:
    min=i
print(min)

-8


# Program 52
# Write a Python program to find largest number in a list.

In [ ]:
num=[8,9,90]
max=num[0]
for i in num:
  if i>max:
    max=i
print(max)

90


# Program 53
# Write a Python program to find second largest number in a list

In [ ]:
def second_high(num):
  num=list(set(num))
  num.sort()
  return num[-2]
nums=[6,7,9,90]
second_high(nums)

9

#alternative

In [ ]:
def sec_max(arr):
  sublist=[x for x in arr if x<max(arr)]
  return max(sublist)
list=[8,9,0,7]
sec_max(list)

8

# Program 54
# Write a Python program to find N largest elements from a list.

In [ ]:
def n_largest(list1,n):
  list1.sort(reverse=True)
  return list1[:n]
list1=[89,78,66,56]
n=3
print(n_largest(list1,n))

[89, 78, 66]


# Program 55
# Write a Python program to print even numbers in a list.


In [ ]:
numbers = [67, 690, 56]
result = [num for num in numbers if num % 2 == 0]
print(result)

[690, 56]


# Program 56
# Write a Python program to print odd numbers in a List.


In [ ]:
numbers=[9,6,7,49,37]
result=[num for num in numbers if num%2!=0]
print(result)

[9, 7, 49, 37]


# Program 57
# Write a Python program to Remove empty List from List.

In [ ]:
list_of_lists=[[1,2,3],[],[4,5],[],[6,7,8],[]]
filtered_list=[ i for i in list_of_lists if i]
print(filtered_list)

[[1, 2, 3], [4, 5], [6, 7, 8]]


# Program 58 Write a Python program to Cloning or Copying a list.

In [ ]:
original_list=[1,2,3,4,5]
cloned_list=original_list[:]
print(cloned_list)

[1, 2, 3, 4, 5]


# Program 59
# Write a Python program to Count occurrences of an element in a list.

In [ ]:
def count_occurrences(lst,element):
  count=0
  for item in lst:
    if item==element:
      count+=1
  return count
my_list=[90,7,7,8,8,8]
ele_count=8
result=count_occurrences(my_list,ele_count)
result

3

# Program 60
# Write a Python program to find words which are greater than given length k.

In [ ]:
def words_greater_than_k(sentence,k):
  words=sentence.split()
  result=[word for word in words if len(word)>k]
  return result
input_sentence = "This is a sample sentence to demonstrate the program"
length_k = 5
words = words_greater_than_k(input_sentence, length_k)
words

['sample', 'sentence', 'demonstrate', 'program']

# Program 61
# Write a Python program for removing 𝑖^th character from a string.

In [ ]:
def remove_ith_character(input_string,i):
  if i<=0 or i>=len(input_string):
    return 'invalid index'
  else:
    return input_string[:i]+input_string[i+1:]
string='example'
index_to_remove=2
result=remove_ith_character(string,index_to_remove)
result

'exmple'

# **Program** 62
# Write a Python program to split and join a string

In [ ]:
def split_and_join(input_string):
  word_list=input_string.split()
  joined_string='_'.join(word_list)
  return word_list,joined_string
string='this is sample string'
split_str,join_str=split_and_join(string)
print(split_str)
print(join_str)

['this', 'is', 'sample', 'string']
this_is_sample_string


# Program 63
# Write a Python program to check if a given string is binary string or not

In [ ]:
def is_binary_string(input_string):
  return set(input_string).issubset({'0','1'})
is_binary_string('101')


True

# Program 64
# Write a Python program to find uncommon words from two Strings.


In [ ]:
def find_uncommon_words(str1,str2):
  return set(str1.split())^(set(str2.split()))
string_1 = "hi how are u?"
string_2 = "hi how was she?"
find_uncommon_words(string_1,string_2)



{'are', 'she?', 'u?', 'was'}

#Program 65
#Write a Python program to find all duplicate characters in string.

In [ ]:
def duplicate_chars(input_string):
  duplicate=[]
  seen=set()
  for char in input_string:
    if char in seen:
      duplicate.append(char)
    else:
      seen.add(char)
  return list(set(duplicate))
duplicate_chars("piyush sharma")

['h', 's', 'a']

# Program 66
# Write a Python Program to check if a string contains any special character.


In [ ]:
import re
def contains_special_character(input_string):
  regex=re.compile('[@_!#$%^&*()<>?/\|}{~:]')
  if regex.search(input_string):
    return True
  else:
    return False
contains_special_character('hello!@%')


True

# Program 67
# Write a Python program to Extract Unique dictionary values.


In [ ]:
def extract_values(input_dict):
  unique_val=list(set(input_dict.values()))
  return unique_val
sample_dict={'a':1,'b':2,'c':1,'d':3,'e':2}
unique_val=extract_values(sample_dict)
print(unique_val)

[1, 2, 3]


# Program 68
# Write a Python program to find the sum of all items in a dictionary.


In [ ]:
def sum_dict_values(input_dict):
  total_sum=0
  for value in input_dict.values():
    total_sum+=value
  return total_sum
def main():
  sample_dict={'a':10,'b':20,'c':30,'d':40}
  result=sum_dict_values(sample_dict)
  print(result)
main()

100


# Program 69
# Write a Python program to Merging two Dictionaries.

In [ ]:
def merge_dict(d1,d2):
  merged_dict=d1.copy()
  merged_dict.update(d2)
  return merged_dict
dict_a={'a':1,"b":2}
dict_b={'c':3,'d':4}
result=merge_dict(dict_a,dict_b)
print(result)

{'a': 1, 'b': 2, 'c': 3, 'd': 4}


#Program 70
# Write a Python program to convert key-values list to flat dictionary.

In [ ]:
def convert_to_flat_dict(key_value_list):
  return dict(key_value_list)
key_value_list=[('a',1),('b',2),('c',3),('d',4)]
result=convert_to_flat_dict(key_value_list)
print(result)

{'a': 1, 'b': 2, 'c': 3, 'd': 4}


# Program 71
# Write a Python program to insertion at the beginning in OrderedDict.


In [ ]:
from collections import OrderedDict
ordered_dict=OrderedDict([('b',2),('c',3),('d',4)])
new_item=('a',1)
new_ordered_dict=OrderedDict([new_item])
new_ordered_dict.update(list(ordered_dict.items()))
print(new_ordered_dict)

OrderedDict([('a', 1), ('b', 2), ('c', 3), ('d', 4)])


# 72  Write a Python program to check order of character in string using OrderedDict().


In [127]:
def check_order_of_characters(input_str,pattern):
  pattern_pointer=0
  for char in input_str:
    if char==pattern[pattern_pointer]:
      pattern_pointer+=1
      if pattern_pointer==len(pattern):
        return True
  return False
input_str='hello world'
search_pattern='u'
result=check_order_of_characters(input_str,search_pattern)
if result:
  print('appear')
else:
  print('disappear')

disappear


# Program 73
# Write a Python program to sort Python Dictionaries by Key or Value.

In [121]:
def sort_dict_key(dictionary):
  sort_dict=dict(sorted(dictionary.items()))
  return sort_dict
def sort_value_key(dictionary):
  sorted_dict=dict(sorted(dictionary.items(),key=lambda item:item[1]))
  return sorted_dict
sample_dict = {'apple': 3, 'banana': 1, 'cherry': 2, 'date': 4}
sort_dict_key(sample_dict)

{'apple': 3, 'banana': 1, 'cherry': 2, 'date': 4}

In [122]:
sort_value_key(sample_dict)

{'banana': 1, 'cherry': 2, 'apple': 3, 'date': 4}

# Program 74
# Write a program that calculates and prints the value according to the given formula:
# Following are the fixed values of C and H:
# C is 50. H is 30.
# D is the variable whose values should be input to your program in a commaseparated sequence.
# Example
# Let us assume the following comma separated input sequence is given to the
# program:
# 100,150,180
# The output of the program should be:
# 18,22,24

In [119]:
def calculate_values(input_sentence):
  C=50
  H=30
  result=[]
  values=list(map(int,input_sentence.split(',')))
  for D in values:
    Q=((2*C*D)/H)**0.5
    result.append(int(Q))
  return result
user_input=input("Enter values ")
res_val=calculate_values(user_input)
print(','.join(map(str,res_val)))

Enter values 100,150,180
18,22,24


# Program 75
# Write a program which takes 2 digits, X,Y as input and generates a 2-dimensional
# array. The element value in the i-th row and j-th column of the array should be i*j.
#Note: i=0,1.., X-1; j=0,1,¡Y-1.
#Example
#Suppose the following inputs are given to the program:
#3,5
#Then, the output of the program should be:
#[[0, 0, 0, 0, 0], [0, 1, 2, 3, 4], [0, 2, 4, 6, 8]]

In [115]:
def gen_2d_array(x,y):
  result=[[i*j for j in range(y)] for i in range(x)]
  return result
x,y=map(int,input('Enter two digits').split(','))
res_arr=gen_2d_array(x,y)
print(res_arr)

Enter two digits3,5
[[0, 0, 0, 0, 0], [0, 1, 2, 3, 4], [0, 2, 4, 6, 8]]


# Program 76
# Write a program that accepts a comma separated sequence of words as input and
# prints the words in a comma-separated sequence after sorting them alphabetically.
# Suppose the following input is supplied to the program:
# without,hello,bag,world
# Then, the output should be:
# bag,hello,without,world

In [113]:
def sorted_word(input_sentence):
  words=input_sentence.split(',')
  sorted_word=sorted(words)
  result=' ,'.join(sorted_word)
  return result
user_input=input("Enter word ")
result=sorted_word(user_input)
print(result)

Enter word without,hello,bag,world
bag ,hello ,without ,world


# Program 77
# Write a program that accepts a sequence of whitespace separated words as input
# and prints the words after removing all duplicate words and sorting them
# alphanumerically.
# Suppose the following input is supplied to the program:
# hello world and practice makes perfect and hello world again
# Then, the output should be:
# again and hello makes perfect practice world

In [109]:
def remove_duplicates_and_sort(words):
  unique_words=sorted(set(words.split()))
  return ' '.join(unique_words)
user_input=input("Enter word ")
result=remove_duplicates_and_sort(user_input)
print(result)

Enter word hello world and practice makes perfect and hello world again
again and hello makes perfect practice world


# Program 79
# Write a program that accepts a sentence and calculate the number of letters and
# digits. Suppose the following input is supplied to the program:
# hello world! 123
# Then, the output should be:
# LETTERS 10
# DIGITS 3

In [105]:
def cal_l_d(sentence):
  l_count=0
  d_count=0
  for char in sentence:
    if char.isalpha():
      l_count+=1
    elif char.isdigit():
      d_count+=1
  return l_count,d_count
user_input=input("Enter a string ")
letters,digits=cal_l_d(user_input)
print(letters)
print(digits)



Enter a string hello world! 123
10
3


# Program 80
# A website requires the users to input username and password to register. Write a
# program to check the validity of password input by users. Following are the criteria
# for checking the password:
# 1. At least 1 letter between [a-z]
# 2. At least 1 number between [0-9]
# 1. At least 1 letter between [A-Z]
# 3. At least 1 character from [$#@]
# 4. Minimum length of transaction password: 6
# 5. Maximum length of transaction password: 12
# Your program should accept a sequence of comma separated passwords and will
# check them according to the above criteria. Passwords that match the criteria are to
# be printed, each separated by a comma.
# Example
# If the following passwords are given as input to the program:
# ABd1234@1,a F1#,2w3E*,2We3345
# Then, the output of the program should be:
# ABd1234@1


In [100]:
import re
def is_valid_password(password):
  if 6<=len(password)<=12:
    if re.search("[a-z]",password):
      if re.search("[A-Z]",password):
        if re.search("[0-9]",password):
          if re.search("[$#@]",password):
            return True
  return False
passwords_input=input("Enter password :")
password=passwords_input.split(',')
valid_passwords=[password for password in passwords if is_valid_password(password)]
print(','.join(valid_passwords))


Enter password :ABd1234@1,a F1#,2w3E*,2We3345
ABd1234@1


# Program 81
# Define a class with a generator which can iterate the numbers, which are divisible by
# 7, between a given range 0 and n.


In [87]:
class div:
  def __init__(self,n):
    self.n=n
  def get_seven(self):
    for i in range(self.n+1):
      if i%7==0:
        yield i
n=int(input('Enter number '))
div_seven=div(n)
for num in div_seven.get_seven():
  print(num)

Enter number 50
0
7
14
21
28
35
42
49


# Program 82
# Write a program to compute the frequency of the words from the input. The output
# should output after sorting the key alphanumerically. Suppose the following input is
# supplied to the program:
# New to Python or choosing between Python 2 and Python 3? Read Python 2 or
# Python 3.
# Then, the output should be:
# 2:2
# 3.:1
# 3?:1
# New:1
#Python:5
# Read:1
# and:1
# between:1
# choosing:1
# or:2
# to:1

In [80]:
def calculate_word_frequency(input_sentence):
  words=input_sentence.split()
  word_freq={}
  for word in words:
    word=word.strip('.?,')
    word=word.lower()
    if word in word_freq:
      word_freq[word]+=1
    else:
      word_freq[word]=1
  sorted_word=sorted(word_freq.items())
  return sorted_word
def print_cal_freq(sorted_word):
  for word,count in sorted_word:
    print(f'{word}:{count}')
input_str=input("enter sentence :")
result=calculate_word_frequency(input_str)
print_cal_freq(result)

enter sentence :New to Python or choosing between Python 2 and Python 3? Read Python 2 or Python 3.
2:2
3:2
and:1
between:1
choosing:1
new:1
or:2
python:5
read:1
to:1


# Program 83
# Define a class Person and its two child classes: Male and Female. All classes have a
# method "getGender" which can print "Male" for Male class and "Female" for Female
# class.

In [63]:
class Person:
  def getGender(self):
    return 'unknown'
class Male(Person):
  def getGender(self):
    return 'male'
class Female(Person):
  def getGender(self):
    return 'female'
p=Person()
m=Male()
f=Female()
print(p.getGender())
print(m.getGender())
print(f.getGender())

unknown
male
female


# Program 84
# Please write a program to generate all sentences where subject is in ["I", "You"] and
# verb is in ["Play", "Love"] and the object is in ["Hockey","Football"].

In [61]:
def gen_sentence(subjects,verbs,objects):
  sentences=[]
  for subject in subjects:
    for verb in verbs:
      for obj in objects:
        sentence=f"{subject} {verb} {obj}"
        sentences.append(sentence)
  return sentences
subjects_list = ["I", "You"]
verbs_list = ["Play", "Love"]
objects_list = ["Hockey", "Football"]

result=gen_sentence(subjects_list,verbs_list,objects_list)
for sentence in result:
  print(sentence)

I Play Hockey
I Play Football
I Love Hockey
I Love Football
You Play Hockey
You Play Football
You Love Hockey
You Love Football


# Program 85
# Please write a program to compress and decompress the string "hello world!hello
# world!hello world!hello world!".

In [56]:
import zlib
def compress_string(original_string):
  compressed_data=zlib.compress(original_string.encode('utf-8'))
  return compressed_data
def decompress_string(compressed_data):
  decode_string=zlib.decompress(compressed_data).decode('utf-8')
  return decode_string
input_sentence= "hello world!hello world!hello world!hello world!"
compress_string(input_sentence)
decompress_string(compressed_data)

'hello world!hello world!hello world!hello world!'

# Program 86
# Please write a binary search function which searches an item in a sorted list. The
# function should return the index of element to be searched in the list.


In [38]:
def binary_search(arr,target):
  left,right=0,len(arr)-1
  while left<=right:
    mid=(left+right)//2
    if arr[mid]==target:
      return mid
    elif arr[mid]<target:
      left=mid+1
    else:
      right=mid-1
  return -1
sorted_list=[1, 2, 3, 4, 5, 6, 7, 8, 9]
target_ele=4
result=binary_search(sorted_list,target_ele)
if result!=1:
  print(f'{target_ele} in {result}')

4 in 3


## Program 87
# Please write a program using generator to print the numbers which can be divisible
# by 5 and 7 between 0 and n in comma separated form while n is input by console.
# Example:
# If the following n is given as input to the program:
# 100
# Then, the output of the program should be:
# 0,35,70


In [27]:
def div_num(n):
  for n in range(n+1):
    if n%5==0 and n%7==0:
      yield n
try:
  n=int(input())
  res=div_num(n)
  print(','.join(map(str,res)))
except ValueError:
  print('invalid')

100
0,35,70


# Program 88
# Please write a program using generator to print the even numbers between 0 and n in
# comma separated form while n is input by console.
# Example:
# If the following n is given as input to the program:
# 10
# Then, the output of the program should be:
# 0,2,4,6,8,10

In [25]:
def eve_num(n):
  for n in range(n+1):
    if n%2==0:
      yield n
try:
  n=int(input())
  res=eve_num(n)
  print(','.join(map(str,res)))
except ValueError:
  print('not posible')

10
0,2,4,6,8,10


# Program 89
# The Fibonacci Sequence is computed based on the following formula:
# f(n)=0 if n=0
# f(n)=1 if n=1
# f(n)=f(n-1)+f(n-2) if n>1
# Please write a program using list comprehension to print the Fibonacci Sequence in
# comma separated form with a given n input by console.
# Example:
# If the following n is given as input to the program:
# 8
# Then, the output of the program should be:
# 0,1,1,2,3,5,8,13

In [18]:
def fib(n):
  if n==0:
    return 0
  elif n==1:
    return 1
  else:
    return fib(n-1)+fib(n-2)
n=int(input())
val=[str(fib(x)) for x in range(0,n+1)]
print(','.join(val))


7
0,1,1,2,3,5,8,13


# Program 90
# Assuming that we have some email addresses in the
# "username@companyname.com (mailto:username@companyname.com)" format,
# please write program to print the user name of a given email address. Both user
# names and company names are composed of letters only.
# Example:
# If the following email address is given as input to the program:
# john@google.com (mailto:john@google.com)
# Then, the output of the program should be:
# john

In [8]:
def extract_username(mail):
  parts=mail.split('@')
  if len(parts)==2:
    return parts[0]
  else:
    return 'invalid format'
extract_username('john@google.com')


'john'

# Program 91
# Define a class named Shape and its subclass Square. The Square class has an init
# function which takes a length as argument. Both classes have an area function which
# can print the area of the shape where Shape's area is 0 by default.

In [ ]:
class shape:
  def __init__(self):
    pass
  def area(self):
    return 0
class squre(shape):
  def __init__(self,length):
    super().__init__()
    self.length=length
  def area(self):
    return self.length**2
shape_obj=squre(5)
print(shape_obj.area())

25


# Program 92
# Write a function that stutters a word as if someone is struggling to read it. The first
# two letters are repeated twice with an ellipsis ... and space after each, and then the
# word is pronounced with a question mark ?.
# Examples
# stutter("incredible") ➞ "in... in... incredible?"
# stutter("enthusiastic") ➞ "en... en... enthusiastic?"
# stutter("outstanding") ➞ "ou... ou... outstanding?"
# Hint :- Assume all input is in lower case and at least two characters long.

In [ ]:
def stutter(word):
  return f"{word[:2]}... {word[:2]}... {word}?"
# Examples
print(stutter("incredible"))     # Output: "in... in... incredible?"
print(stutter("enthusiastic"))   # Output: "en... en... enthusiastic?"
print(stutter("outstanding"))

in... in... incredible?
en... en... enthusiastic?
ou... ou... outstanding?


# Program 93
# Create a function that takes an angle in radians and returns the corresponding angle
# in degrees rounded to one decimal place.
# Examples
# radians_to_degrees(1) ➞ 57.3
# radians_to_degrees(20) ➞ 1145.9
# radians_to_degrees(50) ➞ 2864.8


In [ ]:
import math
def radians_to_degrees(r):
  degree=r*(180/math.pi)
  return round(degree,1)
print(radians_to_degrees(1))
print(radians_to_degrees(20))
print(radians_to_degrees(50))

57.3
1145.9
2864.8


# Program 94
# In this challenge, establish if a given integer num is a Curzon number. If 1 plus 2
# elevated to num is exactly divisible by 1 plus 2 multiplied by num, then num is a
# Curzon number.
# Given a non-negative integer num, implement a function that returns True if num is a
# Curzon number, or False otherwise.
# Examples

#is_curzon(5) ➞ True
# 2 ** 5 + 1 = 33
# 2 * 5 + 1 = 11
# 33 is a multiple of 11
#is_curzon(10) ➞ False
# 2 ** 10 + 1 = 1025
# 2 * 10 + 1 = 21
# 1025 is not a multiple of 21
#is_curzon(14) ➞ True
# 2 ** 14 + 1 = 16385
# 2 * 14 + 1 = 29
# 16385 is a multiple of 29
#Curzon Number:
#It is defined based on a specific mathematical relationship involving powers of 2. An integer
#'n' is considered a Curzon number if it satisfies the following condition:
#If (2^n + 1) is divisible by (2n + 1), then 'n' is a Curzon number.
#For example:
#If n = 5: 2^5 + 1 = 33, and 2*5 + 1 = 11. Since 33 is divisible by 11 (33 % 11 = 0), 5 is a
#Curzon number.
#If n = 10: 2^10 + 1 = 1025, and 2*10 + 1 = 21. 1025 is not divisible by 21, so 10 is not a
#Curzon number.
#Curzon numbers are a specific subset of integers with this unique mathematical property.

In [ ]:
def is_curzon(n):
  num=2**n+1
  den=2*n+1
  return num%den==0
print(is_curzon(5))
print(is_curzon(10))
print(is_curzon(14))


True
False
True


# Program 95
# Given the side length x find the area of a hexagon.
# Examples
# area_of_hexagon(1) ➞ 2.6
# area_of_hexagon(2) ➞ 10.4
# area_of_hexagon(3) ➞ 23.4

In [ ]:
import math
def area_of_hexagon(x):
  area=(3*math.sqrt(3)*x**2)/2
  return round(area,1)
print(area_of_hexagon(1))
print(area_of_hexagon(2))
print(area_of_hexagon(3))

2.6
10.4
23.4


# Program 96
# Create a function that returns a base-2 (binary) representation of a base-10 (decimal)
# string number. To convert is simple: ((2) means base-2 and (10) means base-10)
# 010101001(2) = 1 + 8 + 32 + 128.
# Going from right to left, the value of the most right bit is 1, now from that every bit to
# the left will be x2 the value, value of an 8 bit binary numbers are (256, 128, 64, 32, 16,
# 8, 4, 2, 1).
# Examples
# binary(1) ➞ "1"
# 1*1 = 1
# binary(5) ➞ "101"
# 1*1 + 1*4 = 5
# binary(10) ➞ 1010
# 1*2 + 1*8 = 10

In [ ]:
def binary(decimal):
  if decimal==0:
    return "0"
  binary_rep=""
  while decimal>0:
    rem=decimal%2
    binary_rep=str(rem)+binary_rep
    decimal//=2
  return binary_rep
print(binary(1))    # Output: "1"
print(binary(5))    # Output: "101"
print(binary(10))   # Output: "1010"

1
101
1010


#Program 97
#Create a function that takes three arguments a, b, c and returns the sum of the
#numbers that are evenly divided by c from the range a, b inclusive.
#Examples
#evenly_divisible(1, 10, 20) ➞ 0
# No number between 1 and 10 can be evenly divided by 20.
#evenly_divisible(1, 10, 2) ➞ 30
# 2 + 4 + 6 + 8 + 10 = 30
#evenly_divisible(1, 10, 3) ➞ 18
# 3 + 6 + 9 = 18


In [ ]:
def evenly_divisible(a,b,c):
  total=0
  for i in range(a,b+1):
    if i%c==0:
      total+=i
  return total
print(evenly_divisible(1, 10, 20))
print(evenly_divisible(1, 10, 2))
print(evenly_divisible(1, 10, 3))


0
30
18


# Program 98
# Create a function that returns True if a given inequality expression is correct and
# False otherwise.
# Examples
# correct_signs("3 < 7 < 11") ➞ True
# correct_signs("13 > 44 > 33 <1") ➞ False
# correct_signs("1 < 2 < 6 < 9 > 3") ➞ True



In [ ]:
def correct_signs(expression):
  try:
    return eval(expression)
  except:
    return False
correct_signs("3 < 7 < 11")

True

In [ ]:
correct_signs("13 > 44 > 33 <1")

False

In [ ]:
correct_signs("1 < 2 < 6 < 9 > 3")

True

# Program 99
# Create a function that replaces all the vowels in a string with a specified character.
# Examples
# replace_vowels("the aardvark", "#") ➞ "th# ##rdv#rk"
# replace_vowels("minnie mouse", "?") ➞ "m?nn?? m??s?"
# replace_vowels("shakespeare", "*") ➞ "shkspr*"

In [ ]:
def replace_vowels(string,char):
  vowels='aeiouAEIOU'
  for vowel in vowels:
    string=string.replace(vowel,char)
  return string
print(replace_vowels("the aardvark", "#"))
print(replace_vowels("minnie mouse", "?"))
print(replace_vowels("shakespeare", "*"))

th# ##rdv#rk
m?nn?? m??s?
sh*k*sp**r*


# Program 100
# Write a function that calculates the factorial of a number recursively.
# Examples
# factorial(5) ➞ 120

In [ ]:
def factorial(n):
  if n==0:
    return 1
  else:
    return n*factorial(n-1)
factorial(5)

120

# Program 101
# Hamming distance is the number of characters that differ between two strings.
# To illustrate:
# String1: "abcbba"
# String2: "abcbda"
# Hamming Distance: 1 - "b" vs. "d" is the only difference.
# Create a function that computes the hamming distance between two strings.
# Examples
# hamming_distance("abcde", "bcdef") ➞ 5
# hamming_distance("abcde", "abcde") ➞ 0
# hamming_distance("strong", "strung") ➞ 1

In [ ]:
def hamming_distance(str1,str2):
  if len(str1)!=len(str2):
    raise ValueError('length should be same')
  distance=0
  for i in range(len(str1)):
    if str1[i]!=str2[i]:
      distance+=1
  return distance
print(hamming_distance("abcde", "bcdef"))
print(hamming_distance("abcde", "abcde"))
print(hamming_distance("strong", "strung"))

5
0
1


#Program 102
#Create a function that takes a list of non-negative integers and strings and return a
#new list without the strings.
#Examples
#filter_list([1, 2, "a", "b"]) ➞ [1, 2]
#filter_list([1, "a", "b", 0, 15]) ➞ [1, 0, 15]
#filter_list([1, 2, "aasf", "1", "123", 123]) ➞ [1, 2, 123]

In [ ]:
def filter_list(lst):
  res=[]
  for ele in lst:
    if isinstance(ele,int) and ele>=0:
      res.append(ele)
  return res
filter_list([1, 2, "a", "b"])

[1, 2]

#Program 103
#The "Reverser" takes a string as input and returns that string in reverse order, with
#the opposite case.
##reverse("Hello World") ➞ "DLROw OLLEh"
##reverse("Radar") ➞ "RADAr"

In [ ]:
def reverse(input_string):
  rev_str=input_string[::-1].swapcase()
  return rev_str
reverse("Radar")
reverse("Hello World")

'DLROw OLLEh'

# Program 104
# You can assign variables from lists like this:
# lst = [1, 2, 3, 4, 5, 6]
# first = lst[0]
# middle = lst[1:-1]
# last = lst[-1]
# print(first) ➞ outputs 1
# print(middle) ➞ outputs [2, 3, 4, 5]
# print(last) ➞ outputs 6

In [ ]:
lst = [1, 2, 3, 4, 5, 6]
first,*middle,last=lst
last

6

# Program 105
# Write a function that calculates the factorial of a number recursively.
# Examples
# factorial(5) ➞ 120
# factorial(3) ➞ 6
# factorial(1) ➞ 1
# factorial(0) ➞ 1

In [ ]:
def factorial(n):
  if n==0:
    return 1
  else:
    return n*factorial(n-1)
factorial(3)


6

#Program 106
#Write a function that moves all elements of one type to the end of the list.
#Examples
#move_to_end([1, 3, 2, 4, 4, 1], 1) ➞ [3, 2, 4, 4, 1, 1]
#Move all the 1s to the end of the array.
#move_to_end([7, 8, 9, 1, 2, 3, 4], 9) ➞ [7, 8, 1, 2, 3, 4, 9]
#move_to_end(["a", "a", "a", "b"], "a") ➞ ["b", "a", "a", "a"]


In [ ]:
def move_to_end(lst,ele):
  count=lst.count(ele)
  lst=[x for x in lst if x!=ele]
  lst.extend([ele]*count)
  return lst
move_to_end(["a", "a", "a", "b"], "a")



['b', 'a', 'a', 'a']

## Program 107
# Question1
# Create a function that takes a string and returns a string in which each character is
# repeated once.
# Examples
# double_char("String") ➞ "SSttrriinngg"
# double_char("Hello World!") ➞ "HHeelllloo WWoorrlldd!!"
# double char("1234! ") ➞ "11223344!! "


In [ ]:
def double_char(input_str):
  double_char=""
  for char in input_str:
    double_char+=char*2
  return double_char
double_char("Hello World!")

'HHeelllloo  WWoorrlldd!!'

# Program 108
# Create a function that reverses a boolean value and returns the string "boolean
# expected" if another variable type is given.
# Examples
# reverse(True) ➞ False
# reverse(False) ➞ True
# reverse(0) ➞ "boolean expected"
# reverse(None) ➞ "boolean expected"

In [ ]:
def reverse(value):
  if isinstance(value,bool):
    return not value
  else:
    return "boolean expected"
reverse(0)

'boolean expected'

#Program 109
#Create a function that returns the thickness (in meters) of a piece of paper after
#folding it n number of times. The paper starts off with a thickness of 0.5mm.
#Examples
#num_layers(1) ➞ "0.001m"
#- Paper folded once is 1mm (equal to 0.001m)
#num_layers(4) ➞ "0.008m"
#- Paper folded 4 times is 8mm (equal to 0.008m)
#num_layers(21) ➞ "1048.576m"
#- Paper folded 21 times is 1048576mm (equal to 1048.576m)


In [ ]:
def num_layers(n):
  ini_thick=0.5
  fin_thick=ini_thick*(2**n)
  fin_thick_m=fin_thick/1000
  return fin_thick_m
num_layers(21)

1048.576

# Program 110
# Create a function that takes a single string as argument and returns an ordered list
# containing the indices of all capital letters in the string.
# Examples
# index_of_caps("eDaBiT") ➞ [1, 3, 5]
# index_of_caps("eQuINoX") ➞ [1, 3, 4, 6]
# index_of_caps("determine") ➞ []
# index_of_caps("STRIKE") ➞ [0, 1, 2, 3, 4, 5]
# index_of_caps("sUn") ➞ [1]


In [ ]:
def index_of_caps(word):
  return[i for i,char in enumerate(word) if char.isupper()]
index_of_caps("STRIKE")

[0, 1, 2, 3, 4, 5]

# Program 111
# Using list comprehensions, create a function that finds all even numbers from 1 to
# the given number.
# Examples
#find_even_nums(8) ➞ [2, 4, 6, 8]
#find_even_nums(4) ➞ [2, 4]
#find_even_nums(2) ➞ [2]


In [ ]:
def find_even_nums(num):
  return[x for x in range(1,num+1)if x%2==0]
find_even_nums(8)

[2, 4, 6, 8]

# Program 112
# Create a function that takes a list of strings and integers, and filters out the list so
# that it returns a list of integers only.
# Examples
# filter_list([1, 2, 3, "a", "b", 4]) ➞ [1, 2, 3, 4]
# filter_list(["A", 0, "Edabit", 1729, "Python", 1729]) ➞ [0, 1729]
# filter_list(["Nothing", "here"]) ➞ []

In [ ]:
def filter_list(lst):
  return [x for x in lst if isinstance(x,int)]
filter_list(["A", 0, "Edabit", 1729, "Python", 1729])
filter_list(["Nothing", "here"])
filter_list([1, 2, 3, "a", "b", 4])

[1, 2, 3, 4]

# Program 113
# Given a list of numbers, create a function which returns the list but with each
# element's index in the list added to itself. This means you add 0 to the number at
# index 0, add 1 to the number at index 1, etc...
# Examples
# add_indexes([0, 0, 0, 0, 0]) ➞ [0, 1, 2, 3, 4]
# add_indexes([1, 2, 3, 4, 5]) ➞ [1, 3, 5, 7, 9]
# add_indexes([5, 4, 3, 2, 1]) ➞ [5, 5, 5, 5, 5]


In [ ]:
def add_indexes(lst):
  return [i+v for i,v in enumerate(lst)]
add_indexes([1, 2, 3, 4, 5])

[1, 3, 5, 7, 9]

# Program 114
# Create a function that takes the height and radius of a cone as arguments and returns
# the volume of the cone rounded to the nearest hundredth. See the resources tab for
# the formula.
# Examples
# cone_volume(3, 2) ➞ 12.57
# cone_volume(15, 6) ➞ 565.49
# cone_volume(18, 0) ➞ 0


In [ ]:
import math
def cone_volume(h,r):
  if r==0:
    return 0
  v=1/3*math.pi*(r**2)*h
  return round(v,2)
cone_volume(15, 6)

565.49

#Program 115
#This Triangular Number Sequence is generated from a pattern of dots that form a
#triangle. The first 5 numbers of the sequence, or dots, are:
#1, 3, 6, 10, 15
#This means that the first triangle has just one dot, the second one has three dots, the
#third one has 6 dots and so on.
#Write a function that gives the number of dots with its corresponding triangle number
#of the sequence.
#Examples
#triangle(1) ➞ 1
#triangle(6) ➞ 21
#triangle(215) ➞ 23220

In [ ]:
def triangle(n):
  if n<0:
    return 1
  return n*(n+1)//2
triangle(6)

21

#Program 116
#Create a function that takes a list of numbers between 1 and 10 (excluding one
# number) and returns the missing number.
# Examples
# missing_num([1, 2, 3, 4, 6, 7, 8, 9, 10]) ➞ 5
# missing_num([7, 2, 3, 6, 5, 9, 1, 4, 8]) ➞ 10
# missing_num([10, 5, 1, 2, 4, 6, 8, 3, 9]) ➞ 7

In [ ]:
def missing_num(lst):
  total_list=sum(range(1,11))
  given_list=sum(lst)
  missing_num=total_list-given_list
  return missing_num
missing_num([1, 2, 3, 4, 6, 7, 8, 9, 10])

5

# Program 117
# Write a function that takes a list and a number as arguments. Add the number to the
# end of the list, then remove the first element of the list. The function should then
# return the updated list.
# Examples
# next_in_line([5, 6, 7, 8, 9], 1) ➞ [6, 7, 8, 9, 1]
# next_in_line([7, 6, 3, 23, 17], 10) ➞ [6, 3, 23, 17, 10]
# next_in_line([1, 10, 20, 42 ], 6) ➞ [10, 20, 42, 6]
# next_in_line([], 6) ➞ "No list has been selected"

In [ ]:
def next_in_line(lst,num):
  if lst:
    lst.pop(0)
    lst.append(num)
    return lst
  else:
    return "No list has been selected"
next_in_line([1, 10, 20, 42 ], 6)

[10, 20, 42, 6]

# Program 118
# Create the function that takes a list of dictionaries and returns the sum of people's
# budgets.
# Examples
# get_budgets([
# { 'name': 'John', 'age': 21, 'budget': 23000 },
# { 'name': 'Steve', 'age': 32, 'budget': 40000 },
# { 'name': 'Martin', 'age': 16, 'budget': 2700 }
# ]) ➞ 65700
# get_budgets([
# { 'name': 'John', 'age': 21, 'budget': 29000 },
# { 'name': 'Steve', 'age': 32, 'budget': 32000 },
# { 'name': 'Martin', 'age': 16, 'budget': 1600 }
# ]) ➞ 62600

In [ ]:
def get_budgets(lst):
  bug=sum(person['budget'] for person in lst)
  return bug
get_budgets([
{ 'name': 'John', 'age': 21, 'budget': 29000 },
{ 'name': 'Steve', 'age': 32, 'budget': 32000 },
{ 'name': 'Martin', 'age': 16, 'budget': 1600 }
])

62600

# Program 119
# Create a function that takes a string and returns a string with its letters in
# alphabetical order.
# Examples
#alphabet_soup("hello") ➞ "ehllo"
#alphabet_soup("edabit") ➞ "abdeit"
#alphabet_soup("hacker") ➞ "acehkr"
#alphabet_soup("geek") ➞ "eegk"
#alphabet_soup("javascript") ➞ "aacijprstv"


In [ ]:
def alphabet_soup(letter):
  return ''.join(sorted(letter))
alphabet_soup("javascript")

'aacijprstv'

#Program 120
#Suppose that you invest $10,000 for 10 years at an interest rate of 6% compounded
#monthly. What will be the value of your investment at the end of the 10 year period?
#Create a function that accepts the principal p, the term in years t, the interest rate r,
#and the number of compounding periods per year n. The function returns the value at
#the end of term rounded to the nearest cent.
#For the example:
#compound_interest(10000, 10, 0.06, 12) ➞ 18193.97
#Note that the interest rate is given as a decimal and n=12 because with monthly
#compounding there are 12 periods per year. Compounding can also be done
#annually, quarterly, weekly, or daily.
#Examples
#compound_interest(100, 1, 0.05, 1) ➞ 105.0
#compound_interest(3500, 15, 0.1, 4) ➞ 15399.26
#compound_interest(100000, 20, 0.15, 365) ➞ 2007316.26


In [ ]:
def compound_interest(p,t,r,n):
  a=p*(1+(r/n))**(n*t)
  return round(a,2)
compound_interest(100000, 20, 0.15, 365)


2007316.26

#Program 121
#Write a function that takes a list of elements and returns only the integers.
#Examples
#return_only_integer([9, 2, "space", "car", "lion", 16]) ➞ [9, 2, 16]
#return_only_integer(["hello", 81, "basketball", 123, "fox"]) ➞ [81, 123]
#return_only_integer([10, "121", 56, 20, "car", 3, "lion"]) ➞ [10, 56, 20,3]
#return_only_integer(["String", True, 3.3, 1]) ➞ [1]

In [ ]:
def return_only_integer(lst):
  return [x for x in lst if isinstance(x,int)]
# Examples
print(return_only_integer([9, 2, "space", "car", "lion", 16]))  # Output: [9, 2, 16]
print(return_only_integer(["hello", 81, "basketball", 123, "fox"]))  # Output: [81, 123]
print(return_only_integer([10, "121", 56, 20, "car", 3, "lion"]))  # Output: [10, 56, 20, 3]




[9, 2, 16]
[81, 123]
[10, 56, 20, 3]


### Program 122
# Create a function that takes three parameters where:
# - x is the start of the range (inclusive).
# - y is the end of the range (inclusive).
# - n is the divisor to be checked against.
# Return an ordered list with numbers in the range that are divisible by the third
# parameter n.
#Return an empty list if there are no numbers that are divisible by n.
#Examples
#list_operation(1, 10, 3) ➞ [3, 6, 9]
#list_operation(7, 9, 2) ➞ [8]
#list_operation(15, 20, 7) ➞ []

In [ ]:
def list_operation(x,y,n):
  return[num for num in range(x,y+1)if num%n==0]
list_operation(1, 10, 3)

[3, 6, 9]

# Program 123
# Create a function that takes in two lists and returns True if the second list follows the
# first list by one element, and False otherwise. In other words, determine if the second
# list is the first list shifted to the right by 1.
# Examples
# simon_says([1, 2], [5, 1]) ➞ True
# simon_says([1, 2], [5, 5]) ➞ False
# simon_says([1, 2, 3, 4, 5], [0, 1, 2, 3, 4]) ➞ True
# simon_says([1, 2, 3, 4, 5], [5, 5, 1, 2, 3]) ➞ False
# Notes:
# - Both input lists will be of the same length, and will have a minimum length of 2.
# - The values of the 0-indexed element in the second list and the n-1th indexed
# element in the first list do not matter.

In [ ]:
def simon_says(list1,list2):
  return list1[:-1]==list2[1:]
simon_says([1, 2, 3, 4, 5], [5, 5, 1, 2, 3])

False

In [ ]:
simon_says([1, 2, 3, 4, 5], [0, 1, 2, 3, 4])

True

#Program 124
#A group of friends have decided to start a secret society. The name will be the first
#letter of each of their names, sorted in alphabetical order. Create a function that takes
#in a list of names and returns the name of the secret society.
#Examples
#society_name(["Adam", "Sarah", "Malcolm"]) ➞ "AMS"
#society_name(["Harry", "Newt", "Luna", "Cho"]) ➞ "CHLN"
#society_name(["Phoebe", "Chandler", "Rachel", "Ross", "Monica", "Joey"])

In [ ]:
def society_name(names):
 # Extract the first letter of each name, sort them, and join into a
 secret_name = ''.join(sorted([name[0] for name in names]))
 return secret_name

In [ ]:
society_name(["Phoebe", "Chandler", "Rachel", "Ross", "Monica", "Joey"])

'CJMPRR'

# Program 125
# An isogram is a word that has no duplicate letters. Create a function that takes a
# string and returns either True or False depending on whether or not it's an "isogram".
# Examples

# is_isogram("Algorism") ➞ True
# is_isogram("PasSword") ➞ False
# - Not case sensitive.
# is_isogram("Consecutive") ➞ False
# Notes
# Ignore letter case (should not be case sensitive).
# All test cases contain valid one word strings.

In [ ]:
def is_isogram(word):
  word=word.lower()
  unique_words=set()
  for letter in word:
    if letter in unique_words:
      return False
    unique_words.add(letter)
  return True
is_isogram("Consecutive")


False

# Program 126
# Create a function that takes a string and returns True or False, depending on whether
# the characters are in order or not.
# Examples
# is_in_order("abc") ➞ True
# is_in_order("edabit") ➞ False
# is_in_order("123") ➞ True
# is_in_order("xyzz") ➞ True

In [ ]:
def is_in_order(s):
  return s==''.join(sorted(s))
is_in_order("xyzz")

True

# Program 127
# Create a function that takes a number as an argument and returns True or False
# depending on whether the number is symmetrical or not. A number is symmetrical
# when it is the same as its reverse.
# Examples
# is_symmetrical(7227) ➞ True
# is_symmetrical(12567) ➞ False
# is_symmetrical(44444444) ➞ True
# is_symmetrical(9939) ➞ False
# is_symmetrical(1112111) ➞ True


In [ ]:
def is_symmetrical(num):
  num_str=str(num)
  return num_str==num_str[::-1]
is_symmetrical(1112111)

True

# Program 128
# Given a string of numbers separated by a comma and space, return the product of
# the numbers.
# Examples
# multiply_nums("2, 3") ➞ 6
# multiply_nums("1, 2, 3, 4") ➞ 24
# multiply_nums("54, 75, 453, 0") ➞ 0
# multiply_nums("10, -2") ➞ -20


In [ ]:
def multiply_nums(num_str):
  nums=[int(num)for num in num_str.split(",")]
  result=1
  for num in nums:
    result*=num
  return result
multiply_nums("1, 2, 3, 4")

24

# Program 129
# Create a function that squares every digit of a number.
# Examples
# square_digits(9119) ➞ 811181
# square_digits(2483) ➞ 416649
# square_digits(3212) ➞ 9414
# Notes
# The function receives an integer and must return an integer.

In [ ]:
def square_digits(n):
  n_str=str(n)
  res_str=""
  for digit in n_str:
    square_digits=int(digit)**2
    res_str+=str(square_digits)
  return int(res_str)
square_digits(9119)

811181

# Program 130
# Create a function that sorts a list and removes all duplicate items from it.
# Examples
# setify([1, 3, 3, 5, 5]) ➞ [1, 3, 5]
# setify([4, 4, 4, 4]) ➞ [4]
# setify([5, 7, 8, 9, 10, 15]) ➞ [5, 7, 8, 9, 10, 15]
**setify**([3, 3, 3, 2, 1]) ➞ [1, 2, 3]

In [ ]:
def setify(lst):
  unique_set=set(sorted(lst))
  return list(unique_set)
setify([1, 3, 3, 5, 5])

[1, 3, 5]

# Program 131
# Create a function that returns the mean of all digits.
# Examples
# mean(42) ➞ 3
# mean(12345) ➞ 3
# mean(666) ➞ 6
# Notes
# The mean of all digits is the sum of digits / how many digits there are (e.g. mean
# of digits in 512 is (5+1+2)/3(number of digits) = 8/3=2).
# The mean will always be an integer.


In [ ]:
def mean(n):
  n_str=str(n)
  sum_digit=sum(int(digit) for digit in n_str)
  digit_count=len(n_str)
  return int(sum_digit/digit_count)
mean(666)


6

# Program 132
# Create a function that takes an integer and returns a list from 1 to the given number,
# where:
# 1. If the number can be divided evenly by 4, amplify it by 10 (i.e. return 10 times the
# number).
# 2. If the number cannot be divided evenly by 4, simply return the number.
# Examples

#amplify(4) ➞ [1, 2, 3, 40]
#amplify(3) ➞ [1, 2, 3]
#amplify(25) ➞ [1, 2, 3, 40, 5, 6, 7, 80, 9, 10, 11, 120, 13, 14, 15, 160, 17, 18, 19, 200, 21,
#22, 23, 240, 25]
#Notes
#The given integer will always be equal to or greater than 1.
#Include the number (see example above).
#To perform this problem with its intended purpose, try doing it with list
# comprehensions. If that's too difficult, just solve the challenge any way you can.


In [ ]:
def amplify(num):
  return [n*10 if n%4==0 else n for n in range(1,num+1)]
amplify(25)

[1,
 2,
 3,
 40,
 5,
 6,
 7,
 80,
 9,
 10,
 11,
 120,
 13,
 14,
 15,
 160,
 17,
 18,
 19,
 200,
 21,
 22,
 23,
 240,
 25]

# Program 133
# Create a function that takes a list of numbers and return the number that's unique.
# Examples
# unique([3, 3, 3, 7, 3, 3]) ➞ 7
# unique([0, 0, 0.77, 0, 0]) ➞ 0.77
# unique([0, 1, 1, 1, 1, 1, 1, 1]) ➞ 0
# Notes
# Test cases will always have exactly one unique number while all others are the same.

In [ ]:
def unique(numbers):
  digit_count={}
  for number in numbers:
    if number in digit_count:
      digit_count[number]+=1
    else:
      digit_count[number]=1
  for num,count in digit_count.items():
    if count==1:
      return num
unique([0, 1, 1, 1, 1, 1, 1, 1])

0

## Program 134
# Your task is to create a Circle constructor that creates a circle with a radius provided
# by an argument. The circles constructed must have two getters getArea() (PIr^2) and
# getPerimeter() (2PI*r) which give both respective areas and perimeter
# (circumference).
# For help with this class, I have provided you with a Rectangle constructor which you
# can use as a base example.
# Examples
# circy = Circle(11)
# circy.getArea()
#Should return 380.132711084365
#circy = Circle(4.44)
#circy.getPerimeter()
#Should return 27.897342763877365
#Notes
#Round results up to the nearest integer.


In [ ]:
import math
class circle:
  def __init__(self,radius):
    self.radius=radius
  def getarea(self):
    return (math.pi*self.radius**2)
  def getperimeter(self):
    return (2*math.pi*self.radius)
c=circle(4.44)
c.getarea()
c.getperimeter()


27.897342763877365

# Program 135
# Create a function that takes a list of strings and return a list, sorted from shortest to
# longest.
# Examples
# sort_by_length(["Google", "Apple", "Microsoft"]) ➞ ["Apple", "Google", "Microsoft"]
# sort_by_length(["Leonardo", "Michelangelo", "Raphael", "Donatello"]) ➞ ["Raphael",
# "Leonardo", "Donatello", "Michelangelo"]
# sort_by_length(["Turing", "Einstein", "Jung"]) ➞ ["Jung", "Turing", "Einstein"]
# Notes


In [ ]:
def sort_by_length(strings):
  return sorted(strings,key=len)
sort_by_length(["Turing", "Einstein", "Jung"])

['Jung', 'Turing', 'Einstein']

# Program 136
# Create a function that validates whether three given integers form a Pythagorean
# triplet. The sum of the squares of the two smallest integers must equal the square of
# the largest number to be validated.
# Examples
# is_triplet(3, 4, 5) ➞ True
# 3² + 4² = 25
# 5² = 25
# is_triplet(13, 5, 12) ➞ True
# 5² + 12² = 169
# 13² = 169
# is_triplet(1, 2, 3) ➞ False
# 1² + 2² = 5
# 3² = 9
# Notes
# Numbers may not be given in a sorted order.


In [ ]:
def is_trip(a,b,c):
  num=[a,b,c]
  num.sort()
  return num[0]**2+num[1]**2==num[2]**2
is_trip(13,5,12)

True

# Program 137
# Create a function that takes three integer arguments (a, b, c) and returns the amount
# of integers which are of equal value.
# Examples
# equal(3, 4, 3) ➞ 2
# equal(1, 1, 1) ➞ 3
# equal(3, 4, 1) ➞ 0
# Notes
# Your function must return 0, 2 or 3.


In [ ]:
def equal(a,b,c):
  if a==b==c:
    return 3
  elif a==b or a==c:
    return 2
  else:
    return 0
equal(3,4,3)

2

# Program 138
# Write a function that converts a dictionary into a list of keys-values tuples.
# Examples
# dict_to_list({
# "D": 1,
# "B": 2
# "C": 3
# }) ➞ [("B", 2), ("C", 3), ("D", 1)]
# dict_to_list({
# "likes": 2,
# "dislikes": 3,
# "followers": 10
# }) ➞ [("dislikes", 3), ("followers", 10), ("likes", 2)]
# Notes
# Return the elements in the list in alphabetical order

In [ ]:
def dict_to_list(input_dict):
  sorted_dict=sorted(input_dict.items())
  result=[(key,value)for key,value in sorted_dict]
  return result
dict_to_list({
"likes": 2,
"dislikes": 3,
"followers": 10
})

[('dislikes', 3), ('followers', 10), ('likes', 2)]

# Program 139
# Write a function that creates a dictionary with each (key, value) pair being the (lower
# case, upper case) versions of a letter, respectively.
# Examples
# mapping(["p", "s"]) ➞ { "p": "P", "s": "S" }
# mapping(["a", "b", "c"]) ➞ { "a": "A", "b": "B", "c": "C" }
# mapping(["a", "v", "y", "z"]) ➞ { "a": "A", "v": "V", "y": "Y", "z": "Z" }
# Notes
# All of the letters in the input list will always be lowercase.


In [ ]:
def mapping(letters):
  res={}
  for letter in letters:
    res[letter]=letter.upper()
  return res
mapping(["a", "v", "y", "z"])

{'a': 'A', 'v': 'V', 'y': 'Y', 'z': 'Z'}

# Program 140
# Write a function, that replaces all vowels in a string with a specified vowel.
# Examples
# vow_replace("apples and bananas", "u") ➞ "upplus und bununus"
# vow_replace("cheese casserole", "o") ➞ "chooso cossorolo"
# vow_replace("stuffed jalapeno poppers", "e") ➞ "steffed jelepene peppers"
# Notes
# All words will be lowercase. Y is not considered a vowel.

In [ ]:
def vow_replace(input_str,vowel):
  result=''
  vowels='aeiou'
  for char in input_str:
    if char in vowels:
      result+=vowel
    else:
      result+=char
  return result
vow_replace("stuffed jalapeno poppers", "e")

'steffed jelepene peppers'

# Program 141
# Create a function that takes a string as input and capitalizes a letter if its ASCII code
# is even and returns its lower case version if its ASCII code is odd.
# Examples
# ascii_capitalize("to be or not to be!") ➞ "To Be oR NoT To Be!"
# ascii_capitalize("THE LITTLE MERMAID") ➞ "THe LiTTLe meRmaiD"
# ascii_capitalize("Oh what a beautiful morning.") ➞ "oH wHaT a BeauTiFuL
# moRNiNg."

In [ ]:
def asci_capitalize(input_str):
  result=""
  for char in input_str:
    if ord(char)%2==0:
      result+=char.upper()
    else:
      result+=char.lower()
  return result
asci_capitalize("Oh what a beautiful morning.")

'oH wHaT a BeauTiFuL moRNiNg.'